File: J9Ch/src/J_3_Binary_Tree_Divide_Conquer/Optional_11/_474_Lowest_Common_Ancestor_II_Easy.java
package J_3_Binary_Tree_Divide_Conquer.Optional_11;
import java.util.*;
/**
 * 474
 * Lowest Common Ancestor II
 *
 * Created by tianhuizhu on 6/28/17.
 */
public class _474_Lowest_Common_Ancestor_II_Easy {

    /**
     * Definition of ParentTreeNode:
     *
     * class ParentTreeNode {
     *     public ParentTreeNode parent, left, right;
     * }
     */
    class ParentTreeNode {
        public ParentTreeNode parent, left, right;
    }
    public class Solution {
        /**
         * @param root: The root of the tree
         * @param A, B: Two node in the tree
         * @return: The lowest common ancestor of A and B
         */
        public ParentTreeNode lowestCommonAncestorII(ParentTreeNode root,
                                                     ParentTreeNode A,
                                                     ParentTreeNode B) {
            ArrayList<ParentTreeNode> pathA = getPath2Root(A);
            ArrayList<ParentTreeNode> pathB = getPath2Root(B);

            int indexA = pathA.size() - 1;
            int indexB = pathB.size() - 1;

            ParentTreeNode lowestAncestor = null;
            while (indexA >= 0 && indexB >= 0) {
                if (pathA.get(indexA) != pathB.get(indexB)) {
                    break;
                }
                lowestAncestor = pathA.get(indexA);
                indexA--;
                indexB--;
            }

            return lowestAncestor;
        }

        private ArrayList<ParentTreeNode> getPath2Root(ParentTreeNode node) {
            ArrayList<ParentTreeNode> path = new ArrayList<>();
            while (node != null) {
                path.add(node);
                node = node.parent;
            }
            return path;
        }
    }
}

File: JS9Ch/src/J_3_Binary_Tree_Divide_Conquer/Optional_11/python_cpp/_474_Lowest_Common_Ancestor_II_Easy.cpp
/**
 * 474
 * Lowest Common Ancestor II*/


/**
 * Definition of ParentTreeNode:
 * class ParentTreeNode {
 * public:
 *     int val;
 *     ParentTreeNode *parent, *left, *right;
 * }
 */
class Solution {
public:
    /**
     * @param root: The root of the tree
     * @param A, B: Two node in the tree
     * @return: The lowest common ancestor of A and B
     */
    int Atop, Btop, top;
    ParentTreeNode *a[100000], *b[100000], *ans[100000];
    bool find;
    void inorder(ParentTreeNode *node, ParentTreeNode *A, int flag) {
        if (find==true)
            return;
        if (node == NULL)
            return;
        ans[++top] = node;
        if (A == node) {
            find = true;
            if (flag == 0) {
                Atop = top;
                for (int i = 1; i <= top; ++i)
                    a[i] = ans[i];
            } else {
                Btop = top;
                for (int i = 1; i <= top; ++i)
                    b[i] = ans[i];
            }
            return;
        }

        inorder(node->left, A, flag);
        if (find) return;

        inorder(node->right, A, flag);
        if (find) return;

        top --;

    }

    ParentTreeNode *lowestCommonAncestorII(ParentTreeNode *root,
                                           ParentTreeNode *A,
                                           ParentTreeNode *B) {
        // Write your code here
        top = 0; find = false;
        inorder(root, A, 0);

        top = 0; find = false;
        inorder(root, B, 1);

        Atop = min(Atop, Btop);
        Btop = Atop;

        while (a[Atop] != b[Btop]) {
            Atop --;
            Btop --;
        }
        return a[Atop];
    }
};